Initial Implementation of MediaTek Backend for Executorch

neuropilot-captain opened the pull request:
This pull request introduces the initial implementation of the MediaTek backend for Executorch. Below are the key points and instructions related to this contribution:

- "Ahead of Time" and "Runtime" are included in this implementation.
- The build process and execution phase rely on libraries provided by the MediaTek NeuroPilot SDK. For the purpose of verification, these libraries will be made available offline.

### Build Instructions:
To build the MediaTek backend for Executorch, please adhere to the guidelines provided in the `backends/mediatek/scripts/README.md` file within this repository.

### Ahead-Of-Time Compilation
To compile models for MediaTek devices, please follow the instruction on the ExecuTorch documentation.
For Llama compilation, please refer to `examples/mediatek/README.md` for the usage of the export script `shell_scripts/export_llama.sh`

### On-device Execution:
For instructions on how to run an example on MediaTek devices, refer to the documentation in `examples/mediatek/README.md`.

### Llama Example:
A sample Llama runner is built together with MediaTek backend.

neuropilot-captain committed:
MediaTek Neuron ExecuTorch Backend

- Add backends/mediatek & examples/mediatek
- Add runtime: Neuron ExecuTorch Backend
- Add example: mtk_executor_runner

neuropilot-captain committed:
Set NeuronDelegateSetting from compile specs

neuropilot-captain committed:
Add MediaTek Llama runner

kimishpatel commented on backends/mediatek/scripts/README.md:
```diff
@@ -0,0 +1,13 @@
+Download Android NDK and make sure that $ANDROID_NDK is specified as the NDK Path.
+
+3. Download **MediaTek ExercuTorch libraries**
+Download related libraries.
```
What are these and where do you get it from? I would like to repro

kimishpatel commented on examples/mediatek/README.md:
```diff
@@ -0,0 +1,29 @@
+# ExecuTorch Neuron Backend examples
+
+## Directory structure
```
Can you add pre-requisite here as to which phones support MTK chipset? ALso do add "validated on" to indicate which phone or chipset this is validated on

neuropilot-captain replied:
Thanks for your suggestion! We have added the list of supported chips that the examples are validated on. Also, we have refactor this README for better readability.

kimishpatel commented on examples/mediatek/README.md:
```diff
@@ -0,0 +1,29 @@
+Please follow the build step in backends/mediatek/scripts.
+The mtk_executor_runner will be automatically generated.
+
+## Execute
```
How do you generate the pte file?

kimishpatel commented:
I dont see any instructions/code on the ahead of time component

mergennachin commented:
Thanks for your contribution and PR.

Will take a look at it deeply soon. I just kicked off OSS CI jobs. One thing, we have a lint rule. Here's how you can invoke:

https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#coding-style

mergennachin commented:
> The build process and execution phase rely on libraries provided by the MediaTek NeuroPilot SDK. For the purpose of verification, these libraries will be made available offline.

Is there a public link that we can download from MediaTek website? If not, is there a plan to publish it publicly?

swolchok reviewed (commented):
just dropping in with a bunch of small C++ suggestions; feel free to apply them or not

shoumikhin commented on examples/mediatek/executor_runner/mtk_executor_runner.cpp:
```diff
@@ -0,0 +1,204 @@
+
+  // The memory-planned buffers will back the mutable tensors used by the
+  // method. The sizes of these buffers were determined ahead of time during the
+  // memory-planning pasees.
+  //
+  // Each buffer typically corresponds to a different hardware memory bank. Most
+  // mobile environments will only have a single buffer. Some embedded
+  // environments may have more than one for, e.g., slow/large DRAM and
+  // fast/small SRAM, or for memory associated with particular cores.
+  std::vector<std::unique_ptr<uint8_t[]>> planned_buffers; // Owns the memory
+  std::vector<Span<uint8_t>> planned_spans; // Passed to the allocator
+  size_t num_memory_planned_buffers = method_meta->num_memory_planned_buffers();
+  for (size_t id = 0; id < num_memory_planned_buffers; ++id) {
+    // .get() will always succeed because id < num_memory_planned_buffers.
+    size_t buffer_size =
+        static_cast<size_t>(method_meta->memory_planned_buffer_size(id).get());
+    ET_LOG(Info, "Setting up planned buffer %zu, size %zu.", id, buffer_size);
+    planned_buffers.push_back(std::make_unique<uint8_t[]>(buffer_size));
+    planned_spans.push_back({planned_buffers.back().get(), buffer_size});
+  }
+  HierarchicalAllocator planned_memory(
+      {planned_spans.data(), planned_spans.size()});
+
+  // Assemble all of the allocators into the MemoryManager that the Executor
+  // will use.
+  MemoryManager memory_manager(&method_allocator, &planned_memory);
+
+  //
+  // Load the method from the program, using the provided allocators. Running
+  // the method can mutate the memory-planned buffers, so the method should only
+  // be used by a single thread at at time, but it can be reused.
+  //
+
+  Result<Method> method = program->load_method(method_name, &memory_manager);
+  ET_CHECK_MSG(
+      method.ok(),
+      "Loading of method %s failed with status 0x%" PRIx32,
+      method_name,
+      (uint32_t)method.error());
+  ET_LOG(Info, "Method loaded.");
```
Using the module extension we can skip most of this loading code:

```
#include <executorch/extension/module/module.h>

// ...

Then use it as:

const auto result = module.forward(inputs);

```suggestion
```

cccclai replied:
I'd be careful with this - the current default `module` has some default logic, we'd need to visit if they meet the need or may regress the perf.

neuropilot-captain replied:
The current implementation of `mtk_executor_runner` is based on `examples/portable/executor_runner`, and it has passed testing without any issues.
We will look into `executorch/extension/module` to determine if we need to make any modifications to our code in the future.

cccclai reviewed (commented):
Thank you pushing the first PR! Looks great :)

Overall I think `NeuronBackend` and `mtk_executor_runner` are reasonable. It will be great if we can share the expected flow for users to generate the .pte file and run it.

Regarding the llama runner logic, I can see why we need the multimodel loader and the caching logic, but it's not quite clear to me what logic from `.so`is needed here. Maybe we can walk through it.

neuropilot-captain commented on backends/mediatek/runtime/NeuronExecutor.cpp:
```diff
@@ -0,0 +1,159 @@
+
+    // ---------------------------Execution------------------------------------
+    // Run the compiled model against a set of inputs
+    err = NeuronExecution_create(compilation, &execution);
```
No, this function only creates an execution instance for `execution`, not executing it.

cccclai replied:
The comment said "Run the compiled model against a set of inputs", and I thought it was executing the model. If not, can we update the comment to make it clear?

neuropilot-captain commented on backends/mediatek/runtime/NeuronBackend.cpp:
```diff
@@ -0,0 +1,153 @@
+    return mExecutor.Compute() == NEURON_NO_ERROR ? Error::Ok : Error::InvalidState;
+};
+
+int NeuronExecuTorchDelegate::HintNeuronBackend(EValue** args) const {
```
'import forever' is an optimization hint designed for specific models including Llama. We prefer not to automatically apply it to every model because it could leads unknow effects.

cccclai replied:
I see - hmm probably add more comments about it (or maybe just copy paste your response)

neuropilot-captain commented on examples/mediatek/executor_runner/mtk_executor_runner.cpp:
```diff
@@ -0,0 +1,204 @@
+  // fast/small SRAM, or for memory associated with particular cores.
+  std::vector<std::unique_ptr<uint8_t[]>> planned_buffers; // Owns the memory
+  std::vector<Span<uint8_t>> planned_spans; // Passed to the allocator
+  size_t num_memory_planned_buffers = method_meta->num_memory_planned_buffers();
+  for (size_t id = 0; id < num_memory_planned_buffers; ++id) {
```
As answered in the conversation below.
The current implementation of `mtk_executor_runner` is based on `examples/portable/executor_runner,` and we'll look into `executorch/extension/module `to determine if we need to modify the code.

neuropilot-captain committed:
Merge branch 'pytorch:main' into main

neuropilot-captain committed:
Fix typos and correct variable references in MediaTek backend

- Fix typos in NeuronExecutor.cpp.
- Remove duplicate APUWareUtilsLib.h in CMakeLists.txt.
- Update NEURON_BUFFER_ALLOCATOR to NEURON_BUFFER_ALLOCATOR_LIB.
- Clean up mtk_build.sh and set BUCK_PATH.
- Modify ReadSystemProperty signature to take string reference.

neuropilot-captain committed:
Update README.md

- Add MediaTek libraries introduction
- Improve readability

neuropilot-captain committed:
Update README.md

- Add the list of supported devices
- Improve readability

cccclai commented on backends/mediatek/scripts/README.md:
```diff
@@ -1,13 +1,35 @@
+
+### 3. MediaTek ExercuTorch Libraries
+
+Download the following libraries required for MediaTek ExercuTorch:
```
> Download the following libraries required ...

rephrase it to "Download the following libraries required for MediaTek libraries from xx portal (linked to be added)" just so it's clearer where it comes from

neuropilot-captain committed:
Update README.md

- minor wording change

neuropilot-captain commented on backends/mediatek/runtime/include/NeuronBackend.h:
```diff
@@ -0,0 +1,176 @@
+
+  bool mImportForever = false;
+
+  std::string ToRuntimeOption() {
+    if (mHighAddr && mImportForever) {
+      return "--apusys-config \"{ \\\"high_addr\\\": true, \\\"import_forever\\\": true }\"";
+    } else if (mHighAddr) {
+      return "--apusys-config \"{ \\\"high_addr\\\": true }\"";
+    } else if (mImportForever) {
+      return "--apusys-config \"{ \\\"import_forever\\\": true }\"";
+    } else {
+      return "";
+    }
+  }
+};
```
We appreciate your suggestion to return a const char*, as it's only used to call c_str(). However, we've opted for std::string to maintain flexibility for potential future enhancements.

neuropilot-captain committed:
Update README.md

- Fix Buck2 download link

neuropilot-captain committed:
Update README.md

- Add command example

neuropilot-captain committed:
Merge branch 'pytorch:main' into main

neuropilot-captain committed:
Merge branch 'pytorch:main' into main

neuropilot-captain committed:
Refactor NeuronBackend constants and registration

Changed kHighAddrKey and kImportForeverKey from constexpr to extern in NeuronBackend.h.
Defined kHighAddrKey and kImportForeverKey in NeuronBackend.cpp.
Moved backend registration logic to an anonymous namespace in NeuronBackend.cpp.

neuropilot-captain commented on backends/mediatek/runtime/include/NeuronBackend.h:
```diff
@@ -0,0 +1,176 @@
+  bool is_available() const override;
+};
+
+constexpr char kHighAddrKey[] = "HighAddr";
```
Thank you for your feedback. I have moved the definition to the .cpp file and placed it in an anonymous namespace as suggested.

neuropilot-captain commented on backends/mediatek/runtime/NeuronBackend.cpp:
```diff
@@ -0,0 +1,153 @@
+    LogInfo("NeuronBackend", "version %u, input %u, output %u, length %u, payload size: %zu",
+                 Payload.Header.Version, Payload.Header.InputCount, Payload.Header.OutputCount, Payload.Header.DataLen, processed->size());
+
+    auto delegate = std::unique_ptr<NeuronExecuTorchDelegate>(
```
In our design, a NeuronExecuTorchDelegate handles a **partition** of the model on the NPU. This separation ensures modularity and flexibility, making the codebase cleaner and easier to maintain.

neuropilot-captain commented on backends/mediatek/runtime/NeuronBackend.cpp:
```diff
@@ -0,0 +1,153 @@
+        return Error::InvalidState;
+    };
+
+    auto& allocator = GET_NEURON_ALLOCATOR;
```
I apologize if I misunderstood your question. The NEURON_ALLOCATOR is a global instance that allows users to set up shared memory, which can then be utilized by the NeuronBackend.

neuropilot-captain replied:
Thank you for your response and suggestions.

We understand that we can customize the runtime allocator and point the temporary allocator to `GET_NEURON_ALLOCATOR`, so we don't need to use `NEURON_ALLOCATOR` as a global instance. This is indeed a good solution.

However, in our implementation, `NeuronBackend` needs to determine which device's buffer an address belongs to. Therefore, we use an additional API `Find` from `NEURON_ALLOCATOR` to query the backend. The current API in `BackendExecutionContext` does not allow us to achieve this functionality.

cccclai replied:
It seems like it's reverting to the old implementation, is it expected?

neuropilot-captain committed:
Fix comment

mergennachin commented on backends/mediatek/runtime/NeuronBackend.cpp:
```diff
@@ -0,0 +1,162 @@
+
+void NeuronBackend::destroy(DelegateHandle* handle) const {
+    NeuronExecuTorchDelegate* delegate = reinterpret_cast<NeuronExecuTorchDelegate*>(handle);
+    delete delegate;
```
you might have to call the destructor directly. Something like this. `delegate->~NeuronExecuTorchDelegate()`

neuropilot-captain committed:
Add MTK AoT backend and AoT Flow with llama as example

cccclai commented on examples/mediatek/model_export_scripts/llama.py:
```diff
@@ -0,0 +1,184 @@
+        "Model config must be in same directory as all model weight bins and tokenizer files.")
+    parser.add_argument('-p', '--precision', type=str, default='A16W8', choices=['A16W4', 'A16W8', 'A16W16', 'A8W4', 'A8W8'],
+        help="Precision to quantize entire model to.")
+    parser.add_argument('-n', '--num_chunks', type=int, default=4,
```
If it's stories model (the same model arch as llama2, 125M parameters), is the flow going to work? I believe in this case, we don't need to chunk the model: 
model checkpoint: https://huggingface.co/karpathy/tinyllamas/resolve/main/stories110M.pt
tokenizer: https://raw.githubusercontent.com/karpathy/llama2.c/master/tokenizer.model

neuropilot-captain replied:
Hi @cccclai , generally, if its <= 3B param models we can do 1 chunk (ie effectively no chunking). 

Users just need to set -n 1 in the export shell script. 

For >= 7B params we need minimum 2 chunks.

cccclai commented on examples/mediatek/models/llm_models/weights/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*
```
Is it a wip folder?

neuropilot-captain replied:
Mainly just to indicate to users where to put their model weights in.

cccclai commented on examples/mediatek/models/llm_models/modeling_common.py:
```diff
@@ -0,0 +1,497 @@
+        )
+        # Note: Assume cache size fixed shape as torch dynamic shape cannot handle dim 3 being
+        # combination of 2 dynamic dims
+        if get_dym_shape:
```
Are we using dynamic shape?

neuropilot-captain replied:
Dynamic Shape is more for the case where users would like to calibrate on their own dataset. 

With dynamic shape, users can avoid padding each input to the length. 

Also, it is only used during the `capture_pre_autograd_graph`  part of the export flow. 

Subsequently the shape is fixed during the `torch.export.export` part

cccclai commented on examples/mediatek/models/llm_models/modeling_llama.py:
```diff
@@ -0,0 +1,76 @@
+        dtype=torch.float32,
+        include_tail=False,
+        return_attn=False,
+        jit_trace=False,
```
what does `jit_trace` mean?

neuropilot-captain replied:
It's a flag to skip all the forced float casting and some assertion checks.

cccclai reviewed (commented):
Thank you for putting up the AOT part prs! It's a long pr and I'm just half on it. Will continue reviewing later.

neuropilot-captain committed:
Merge branch 'pytorch:main' into main

cccclai commented on examples/mediatek/model_export_scripts/llama.py:
```diff
@@ -0,0 +1,184 @@
+
+    print("Getting pre autograd ATen Dialect Graph")
+    pre_autograd_aten_dialect = torch._export.capture_pre_autograd_graph(model, example_inputs, dynamic_shapes=dynamic_shapes) # NOTE: Will be replaced with export
+    quantizer = NeuropilotQuantizer()
```
Is `NeuropilotQuantizer` hosted on NP portal or github?

cccclai commented on backends/mediatek/neuropilot/preprocess.py:
```diff
@@ -0,0 +1,64 @@
+            value = spec.value.decode('utf-8')
+            compile_options.append(f'--{spec.key}={value}')
+
+        converter = mtk_converter.PyTorchV2Converter.from_exported_program(edge_program)
```
It seems like the `mtk_convert` hosted on NP portal is handling go through the graph and process node by node.

From a user perspective, without knowing what ops can be supported by mtk is a bit hard to lower a model to mtk backend. There can be two scenarios:
1. Partially lower the model. While it is the plan to partially lower some oss models to npu, it's a bit hard to tell if it will be supported smoothly as everything is hidden behind `mkt_convert`. 
2. Decompose the not supported ops to the op list supported by npu to fully lower a model to mtk. Similar to scenario 2 (also similar the `decompose_scaled_dot_product_attention` pass), users may want to decompose some non supported ops with the supported ops. It's hard to do so without know what ops are supported.

cccclai commented on backends/mediatek/neuropilot/quantizer/annotator.py:
```diff
@@ -0,0 +1,263 @@
+
+@register_annotator(
+    [
+        torch.ops.aten.add.Scalar,
+        torch.ops.aten.add.Tensor,
+        torch.ops.aten.add_.Scalar,
+        torch.ops.aten.add_.Tensor,
+        torch.ops.aten.bmm.default,
```
I asked this question in the meeting and actually this part of code answered my question. The quantizer looks reasonable to me now.

cccclai replied:
Yeah, it's more like describing how the ops are quantized.

cccclai reviewed (commented):
I just finish reading through - the process in `examples/mediatek/model_export_scripts/llama.py` looks neat. Thank you for sending the high quality code.

One question I have is mostly regarding the supported ops in the backend.preprocess. Mostly around what ops can be supported by users and how to allow users to figure out by themselves to either partially lower their models to npu or fully lower to npu.

neuropilot-captain committed:
Fix 1t model export bug

neuropilot-captain committed:
Move AOT code to `backends/mediatek` and add skip ops mechanism

neuropilot-captain committed:
Add calibration flow and update compile options

neuropilot-captain committed:
Update llama runner

neuropilot-captain committed:
Update llama runner sample run script

neuropilot-captain committed:
Add embedding bin dumping for cmdline

neuropilot-captain committed:
Upload llama3 8B instruct model config and tokenizer config files for testing

neuropilot-captain committed:
Update README.md

neuropilot-captain committed:
Add `op_names_to_skip` argument to NeuropilotPartitioner

Riandy commented on examples/mediatek/README.md:
- Add that only linux environment is supported (this is due to the extra libs mtk_neuron and mtk_converter whl files are only supported in linux)
- Add hyperlink to ET environment setup https://pytorch.org/executorch/stable/getting-started-setup.html  
- Mention that requirements.txt file has two dependencies that cannot be installed via pip. 
  - mtk_converter
  - mtk-neuron
  - Another option is to split this up by removing those two from requirements.txt, so devs can just `pip3 install -r requirements.txt` and then explicitly mention that they need to install the other two libs.

```
// Ensure that you are inside executorch/examples/mediatek directory
pip3 install -r requirements.txt

// Download the two whl files from (some_location/URL)
pip3 install mtk_neuron-8.2.2-py3-none-linux_x86_64.whl
pip3 install mtk_converter-8.8.0.dev20240723+public.d1467db9-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
```
- Explain that dev should expect 8 pte files (4 for prompt and 4 for generation) + 1 token embedding file and provide location of the token embedding file location 
 ` models/llm_models/weights/<model_name>/<embedding_file_name>.bin `
- For llama models, need to use `mtk_llama_executor_runner`, and ask them to check `examples/mediatek/executor_runner/run_llama3_sample.sh` for reference

neuropilot-captain replied:
> mtk_neuron seems to require a localhost connection. So verify that localhost connection is available (not blocked) before running the AoT script

Do you mean to guide users check their network condition before running this example?

Riandy replied:
Yes. Give them some warning that a working localhost connection is required in AOT flow. When I tested the flow in Windows + Linux WSL (localhost is blocked) and I wasted 1.5 hours on the initial AoT flow without knowing that localhost is required.

Riandy commented on backends/mediatek/scripts/README.md:
- Buck2 Build Tool
  - This seems to be not needed? Looks like ET will auto download the required version it needs? (See: `executorch/build/resolve_buck.py`)
- Android NDK
  - Any specific version that is needed? Or tested upon? Maybe we should specify the version tested?
- “Build Script” section
  - Mentioned the output of the build script. Which is in the “`cmake-android-out`” folder.
In this case, it looks like we have 3 runners generated (executorch_runner, mtk_executor_runner and `mtk_llama_executor_runner`). `mtk_llama_executor_runner` is located inside `/cmake-android-out/examples/mediatek`
  - Question: should `executorch_runner` ever be used? If not, maybe not generate it?
Mention that they should use `mtk_llama_executor_runner` for llama models

neuropilot-captain replied:
No, we should remove in the future

Commit:
Add script to build MediaTek examples.

Commit:
Leverage temp allocator for memory management

neuropilot-captain committed:
Update README and remove mtk_neuron and mtk_converter from requirements.txt

Commit:
Replace delete with destructor in backend destroy

neuropilot-captain committed:
Respect the model input/output data types (quantized or float) in Neuropilot backend

neuropilot-captain committed:
Update README.md

Update for example build script

cccclai commented:
The changes in root cmake and the buffer allocator looks good! I marked them as resolved. The only one left is the destroy function and we're good then.

neuropilot-captain committed:
Fix neuron backend linked library

Link executorch_no_prim_ops instead of executorch

neuropilot-captain committed:
Use temp allocator in neuron backend

cccclai commented:
Many thanks for MTK team for making the contribution. I'll go ahead to merge it. Need to wait for the CI signal a bit.

cccclai commented:
Looks like there are some lint errors. Could you address them? Here is the failing job https://github.com/pytorch/executorch/actions/runs/10296603326/job/28529110807?pr=3571

neuropilot-captain committed:
Add missing module import for `FakeQuantize` class

neuropilot-captain committed:
Fix Python linter error

neuropilot-captain committed:
Update mask_builder.h

Commit:
Fix lint errors of newlines

neuropilot-captain committed:
fix tokenizer lint error

neuropilot-captain committed:
Apply lintrunner patches

neuropilot-captain committed:
fix argument mismatch

neuropilot-captain committed:
Merge branch 'pytorch:main' into main

neuropilot-captain committed:
Merge branch 'pytorch:main' into main
